src/main/java/io/debezium/connector/vitess/connection/VitessReplicationConnection.java

```diff
@@ -89,28 +91,86 @@ public void startStreaming(
         }
 
         StreamObserver<Vtgate.VStreamResponse> responseObserver = new StreamObserver<Vtgate.VStreamResponse>() {
+            private List<VEvent> bufferedEvents = new ArrayList<>();
+            private Vgtid newVgtid;
+            private boolean beginEventSeen;
+            private boolean commitEventSeen;
+            private int numOfRowEvents;
+            private int numResponses;
 
             @Override
             public void onNext(Vtgate.VStreamResponse response) {
-
                 LOGGER.debug("Received {} VEvents in the VStreamResponse:",
                         response.getEventsCount());
-                for (VEvent vEvent : response.getEventsList()) {
-                    LOGGER.debug("VEvent: {}", vEvent);
+                boolean sendNow = false;
+                for (VEvent event : response.getEventsList()) {
+                    LOGGER.debug("VEvent: {}", event);
+                    switch (event.getType()) {
+                        case ROW:
+                            numOfRowEvents++;
+                            break;
+                        case VGTID:
+                            // We always use the latest VGTID if any.
+                            newVgtid = Vgtid.of(event.getVgtid());
+                            break;
+                        case BEGIN:
+                            // We should only see BEGIN before seeing COMMIT.
+                            if (commitEventSeen) {
+                                LOGGER.error("Receive BEGIN event after receiving COMMIT event");
+                            }
+                            if (beginEventSeen) {
+                                LOGGER.error("Received duplicate BEGIN events");
+                            }
+                            beginEventSeen = true;
+                            break;
+                        case COMMIT:
+                            // We should only see COMMIT after seeing BEGIN.
+                            if (!beginEventSeen) {
+                                LOGGER.error("Receive COMMIT event before receiving BEGIN event");
+                            }
+                            if (commitEventSeen) {
+                                LOGGER.error("Received duplicate COMMIT events");
+                            }
+                            commitEventSeen = true;
+                            break;
+                        case DDL:
+                        case OTHER:
```

Comment:
We just process them (send them to) to VStreamOutputMessageDecoder.processMessage, then in tje processor (VitessStreamingChangeEventSource.newReplicationMessageProcessor) it will merely rotate the vgtid for offset. These two events will be not be send as SourceRecords to Kafka.